### PR TITLE
FEATURE: TNT-1590 Column Headers Position Left

### PR DESCRIPTION
**Pull request by NestorEncinas**
---

Supported PR commands:

| Command                                                 | Description                                                |
| ------------------------------------------------------- | ---------------------------------------------------------- |
| `ok to test`                                            | Re-run standard checks                                     |
| `extended check sonar`                                  | SonarQube tests                                            |
| `extended test - backstop`                              | BackstopJS tests                                           |
| **E2E Cypress tests commands - TIGER**                  |                                                            |
| `extended test - tiger-cypress - isolated <testName>`   | Run isolated tests running against recorded Tiger backend. |
| `extended test - tiger-cypress - record <testName>`     | Create a new recording for isolated Tiger tests.           |
| `extended test - tiger-cypress - integrated <testName>` | Run integrated tests against live backend                  |
| **E2E Cypress tests commands - BEAR**                   |                                                            |
| `extended test - cypress - isolated <testName>`         | Run isolated tests running against recorded Bear backend.  |
| `extended test - cypress - record <testName>`           | Create a new recording for isolated Bear tests.            |
| `extended test - cypress - integrated <testName>`       | Run integrated tests against live backend                  |
| **Compatibility matrix test commands - TIGER Backend**  |                                                            |
| `extended test - matrix-test <AIO_version>`             | Run integrated tests against AIO versions.                 |

`<testName>` in cypress commands is used to filter specfiles. Example, to run record with BEAR backend

-   Against `dashboard.spec.ts` and `drilling.spec.ts`, execute command `extended test - cypress - record dashboard,drilling`
-   Against all specfiles, execute command `extended test - cypress - record` or `extended test - cypress - record *`

`<AIO_version>` in commands is used to start test with multiple AIO instances - each instance in triggered by one jenkins build

-   To run with `master` and `stable`, execute command `extended test - matrix-test master,stable` or `extended test - matrix-test latest`
-   To run with specific version,ex: `2.3.0` and `2.3.1`, execute command `extended test - matrix-test 2.3.0,2.3.1`
-   In case `<AIO_version>` is empty, read versions from file `compTigerVersions.txt` of this repo

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `extended test - backstop` passes
-   [ ] `extended test - tiger-cypress - record` to record new mapping files (Tiger BE)
-   [ ] `extended test - cypress - record` to record new mapping files (Bear BE)
-   [ ] `extended test - tiger-cypress - isolated` passes
-   [ ] `extended test - cypress - isolated` passes
-   [ ] `extended test - tiger-cypress - integrated` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14393/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15942/ : FAILURE in 9m 06s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14396/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15946/ : FAILURE in 8m 12s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14640/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16189/ : FAILURE in 8m 52s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14661/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16215/ : FAILURE in 8m 46s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14664/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16216/ : FAILURE in 8m 00s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14681/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16240/ : FAILURE in 8m 01s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14749/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16307/ : FAILURE in 13m 46s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14766/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16324/ : FAILURE in 7m 57s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14774/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16333/ : FAILURE in 13m 42s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14776/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16336/ : FAILURE in 2m 43s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14782/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16340/ : FAILURE in 8m 54s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14787/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16349/ : FAILURE in 19m 12s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14789/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16354/ : FAILURE in 8m 13s

**Comment by ivan-nejezchleb** [comment]
extended test - backstop

**Comment by yenkins** [comment]
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/2355/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14800/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/1119/ : UNSTABLE in 26m 21s

**Comment by ivan-nejezchleb** [comment]
extended test - backstop

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14810/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16377/ : SUCCESS in 11m 55s

**Comment by yenkins** [comment]
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/2356/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by ivan-nejezchleb** [comment]
extended test - backstop

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14818/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16383/ : SUCCESS in 12m 04s

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14821/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/1122/ : FAILURE in 17m 35s

**Comment by ivan-nejezchleb** [comment]
extended test - backstop

**Comment by yenkins** [comment]
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/2359/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14822/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/1123/ : UNSTABLE in 26m 13s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14834/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16397/ : SUCCESS in 13m 07s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14845/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16408/ : SUCCESS in 11m 53s

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14854/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16423/ : FAILURE in 1m 40s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14859/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16429/ : FAILURE in 11m 57s

**Comment by ivan-nejezchleb** [comment]
extended test - backstop

**Comment by yenkins** [comment]
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/2368/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14866/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/1132/ : UNSTABLE in 25m 47s

**Comment by ivan-nejezchleb** [comment]
extended test - backstop

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14867/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16438/ : SUCCESS in 11m 52s

**Comment by yenkins** [comment]
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/2369/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/14868/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/1133/ : SUCCESS in 25m 53s

File: libs/sdk-ui-ext/src/internal/utils/controlsHelper.ts [hunk 1]
```diff
@@ -140,3 +149,14 @@ export function minInputValidateAndPushData(
 
     pushData({ properties, propertiesMeta });
 }
+
+export function isSetColumnHeadersPositionToLeftAllowed(insight: IInsightDefinition) {
+    const rowAttribute = bucketsFind(insightBuckets(insight), BucketNames.ATTRIBUTE);
+    const hasRows = rowAttribute && bucketItems(rowAttribute).length > 0;
+    const columnAttribute = bucketsFind(insightBuckets(insight), BucketNames.COLUMNS);
```
Comment (pbenes) [inline libs/sdk-ui-ext/src/internal/utils/controlsHelper.ts:156]: please rename to columnAttribute -> columnBucket

File: libs/sdk-ui-pivot/src/impl/data/dataSource.ts [hunk 2]
```diff
@@ -97,8 +98,14 @@ export class AgGridDatasource implements IDatasource {
 
         this.grouping.processPage(rowData, offset[0], rowAttributeIds);
 
+        const columnHeadersPosition = this.config.getColumnHeadersPosition();
         // RAIL-1130: Backend returns incorrectly total: [1, N], when count: [0, N] and offset: [0, N]
-        const lastRow = offset[0] === 0 && count[0] === 0 ? 0 : totalCount[0];
+        const lastRow =
```
Comment (pbenes) [inline libs/sdk-ui-pivot/src/impl/data/dataSource.ts:103]: please avoid nested ternary operators

File: libs/sdk-ui-pivot/src/impl/data/rowFactory.ts [hunk 3]
```diff
@@ -324,27 +331,78 @@ export function createAgGridPage(
...
-    };
+            const row: IGridRow = {
+                [headerColumn.id]: attributeName,
+                headerItemMap: {},
+            };
+
+            tableDescriptor.headers.mixedValuesCols.forEach((column, columnIndex) => {
+                const header = attributes[columnIndex];
+                if (isResultAttributeHeader(header)) {
+                    // TODO what about formattedName?
```
Comment (pbenes) [inline libs/sdk-ui-pivot/src/impl/data/rowFactory.ts:351]: what should we do with this TODO?

File: libs/sdk-ui-pivot/src/impl/structure/colDefFactory.ts [hunk 4]
```diff
@@ -282,7 +315,10 @@ export function createColDefsFromTableDescriptor(
 
     createAndAddSliceColDefs(table.sliceCols, table.sliceMeasureCols, state);
     createAndAddDataColDefs(table, state, intl);
-    // handle metrics in rows and no column attribute case
+    if (config?.columnHeadersPosition === "left") {
```
Comment (pbenes) [inline libs/sdk-ui-pivot/src/impl/structure/colDefFactory.ts:318]: on other places, the check also includes the isTransposed() part. Consider adding it here and you may even remove the isTransposed from sending to the method and remove the if statement here https://github.com/gooddata/gooddata-ui-sdk/pull/3865/files#diff-ca12b4453929dd05baa87f0e6b40f136da5fdbacf1d25985af91612288bb3641R133

File: libs/sdk-ui-pivot/src/impl/structure/tableDescriptor.ts [hunk 5]
```diff
@@ -65,6 +73,8 @@ export class TableDescriptor {
     public readonly zippedLeaves: Array<[LeafDataCol, ColDef]> = [];
     private readonly _seriesColsCount: number;
 
+    // TODO solve zipping and lookup (does not seems to be used in the new transpose mode for now)
```
Comment (pbenes) [inline libs/sdk-ui-pivot/src/impl/structure/tableDescriptor.ts:76]: remove?

File: libs/sdk-ui-pivot/src/impl/data/rowFactory.ts (hunk 3 above)
Comment (ivan-nejezchleb) [reply libs/sdk-ui-pivot/src/impl/data/rowFactory.ts:351]: formatted name already used with precedents inside getAttributeHeaderItemName. Removing TODO

File: libs/sdk-ui-pivot/src/impl/structure/tableDescriptor.ts (hunk 5 above)
Comment (ivan-nejezchleb) [reply libs/sdk-ui-pivot/src/impl/structure/tableDescriptor.ts:76]: removed

File: libs/sdk-ui-pivot/src/impl/structure/colDefFactory.ts (hunk 4 above)
Comment (ivan-nejezchleb) [reply libs/sdk-ui-pivot/src/impl/structure/colDefFactory.ts:318]: explained by comment

File: libs/sdk-ui-pivot/src/impl/data/dataSource.ts (hunk 2 above)
Comment (ivan-nejezchleb) [reply libs/sdk-ui-pivot/src/impl/data/dataSource.ts:103]: fixed

File: libs/sdk-ui-ext/src/internal/utils/controlsHelper.ts (hunk 1 above)
Comment (ivan-nejezchleb) [reply libs/sdk-ui-ext/src/internal/utils/controlsHelper.ts:156]: fixed

File: libs/sdk-ui-pivot/src/impl/structure/colDefHeaderClass.ts [hunk 6]
```diff
@@ -125,3 +133,7 @@ function getNoBottomBorderGroupHeader(colDesc: AnyCol, tableDescriptor: TableDes
         return isScopeCol(colDesc) && isResultTotalHeader(colDesc.header);
     }
 }
+
+function hasEmptyChild(colDef: ColGroupDef, colDesc: AnyCol) {
+    return isMixedHeadersCol(colDesc) && colDef.children?.[0] && colDef.children[0].headerName === "";
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply libs/sdk-ui-pivot/src/impl/structure/colDefHeaderClass.ts:138]: fixed

File: libs/sdk-ui-pivot/src/impl/gridOptions.ts [hunk 7]
```diff
@@ -144,5 +150,7 @@ export function createGridOptions(
         rowHeight: DEFAULT_ROW_HEIGHT,
         autoSizePadding: DEFAULT_AUTOSIZE_PADDING,
         enableBrowserTooltips: true,
+
+        headerHeight: hideEmptyHeader ? 0 : undefined,
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply libs/sdk-ui-pivot/src/impl/gridOptions.ts:154]: added

File: libs/sdk-ui-pivot/src/impl/drilling/cellDrillabilityPredicate.ts [hunk 8]
```diff
@@ -15,7 +15,8 @@ export function isCellDrillable(
     dv: DataViewFacade,
     drillablePredicates: IHeaderPredicate[],
 ): boolean {
-    if (drillablePredicates.length === 0) {
+    // TODO
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply libs/sdk-ui-pivot/src/impl/drilling/cellDrillabilityPredicate.ts:18]: removed

File: libs/sdk-ui-pivot/src/impl/tableFacade.ts [hunk 9]
```diff
@@ -881,6 +890,17 @@ export class TableFacade {
...
     public getRowCount = (): number => {
+        if (this.config?.columnHeadersPosition === "left" && this.tableDescriptor.isTransposed()) {
+            /**
+             * in case of metrics in rows and column attributes on left
+             * the table has no header and its elements hierarchy is rendered in normal rows.
+             * It completely messes up ag-grid approach and data loading via dataSource. Only benefit is to have all rows scrollable instead of sticky header.
+             */
+            return (
+                this.visibleData.rawData().firstDimSize() +
+                this.visibleData.meta().dimensions()[1].headers.length
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply libs/sdk-ui-pivot/src/impl/tableFacade.ts:901]: replaced

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Review by pbenes — APPROVED** [review]

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR3865

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/3018/ : SUCCESS in 11m 23s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16462/ : SUCCESS in 11m 47s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-bear-isolated-zuul-pipeline/572/ : SUCCESS in 10m 45s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/4177/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul/899/ : FAILURE in 1m 32s (non-voting)
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-tiger-isolated-zuul-pipeline/525/ : SUCCESS in 7m 58s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-compatibility-pipeline/393/ : SUCCESS in 5m 53s

**Pull request merged**